File: Telephone/CallTransferController.m
//
//  CallTransferController.m
//  Telephone
//
//  Copyright © 2008-2016 Alexey Kuznetsov
//  Copyright © 2016-2022 64 Characters
//
//  Telephone is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//

#import "CallTransferController.h"

#import "AKSIPCall.h"

#import "AccountController.h"
#import "CallController+Protected.h"


@interface CallTransferController ()

@property(nonatomic, weak) CallController *sourceCallController;

@end


@implementation CallTransferController

- (instancetype)initWithSourceCallController:(CallController *)callController
                                   userAgent:(AKSIPUserAgent *)userAgent {
    AccountController *accountController = callController.accountController;

    self = [self initWithWindowNibName:@"CallTransfer"
                    accountController:accountController
                            userAgent:userAgent
                             delegate:accountController];
    if (self == nil) {
        return nil;
    }

    _sourceCallController = callController;
    [self showInitialState:self];
    return self;
}

- (void)transferCall {
    [self.sourceCallController.call attendedTransferToCall:self.call];
}

- (void)closeSheet:(id)sender {
    CallController *sourceCallController = self.sourceCallController;
    if (sourceCallController.isCallActive &&
        sourceCallController.isCallOnHold) {
        [sourceCallController toggleCallHold];
    }

    [sourceCallController dismissCallTransfer];
    [sourceCallController discardCallTransfer];
}

- (void)showInitialState:(id)sender {
    if (self.isCallActive) {
        [self hangUpCall];
    }

    if (!self.sourceCallController.isCallActive) {
        [self closeSheet:self];
        return;
    }

    [self showTransferDestinationState];
    [self focusTransferDestination];
}


#pragma mark - CallController methods

- (CallTransferController *)callTransferController {
    return nil;
}

- (void)acceptCall {
    // Transfer destination calls are outgoing only.
}

- (void)prepareForCall {
    [super prepareForCall];
    [self setTransferActionEnabled:NO];
}


#pragma mark - AKSIPCall notifications

- (void)SIPCallEarly:(NSNotification *)notification {
    [super SIPCallEarly:notification];
    [self setTransferActionEnabled:NO];
}

- (void)SIPCallDidDisconnect:(NSNotification *)notification {
    [super SIPCallDidDisconnect:notification];
}

- (void)SIPCallDidLocalHold:(NSNotification *)notification {
    [super SIPCallDidLocalHold:notification];
    [self callDidHoldForTransfer];
}


@end

File: Telephone/CallTransferController.swift
//
//  CallTransferController.swift
//  Telephone
//
//  Small specialization of CallController for an attended transfer.
//

import Foundation

@MainActor
@objc(CallTransferController)
@objcMembers
final class CallTransferController: CallController {
    private weak var sourceCallController: CallController?

    @objc(initWithSourceCallController:userAgent:)
    init(
        sourceCallController: CallController,
        userAgent: AKSIPUserAgent
    ) {
        self.sourceCallController = sourceCallController

        super.init(
            windowNibName: "CallTransfer",
            accountController: sourceCallController.accountController,
            userAgent: userAgent,
            delegate: sourceCallController.accountController
        )

        showInitialState(nil)
    }

    override var callTransferController: CallTransferController? {
        nil
    }

    func transferCall() {
        guard
            let sourceCall = sourceCallController?.call,
            let destinationCall = call
        else {
            return
        }

        sourceCall.attendedTransfer(to: destinationCall)
    }

    @objc(closeSheet:)
    func closeSheet(_ sender: Any?) {
        guard let sourceCallController else { return }

        if sourceCallController.callActive,
           sourceCallController.callOnHold
        {
            sourceCallController.toggleCallHold()
        }

        sourceCallController.dismissCallTransfer()
        sourceCallController.discardCallTransfer()
    }

    @objc(showInitialState:)
    func showInitialState(_ sender: Any?) {
        if callActive {
            hangUpCall()
        }

        guard sourceCallController?.callActive == true else {
            closeSheet(nil)
            return
        }

        showTransferDestinationState()
        focusTransferDestination()
    }

    override func acceptCall() {
        // Transfer destination calls are outgoing only.
    }

    override func prepareForCall() {
        super.prepareForCall()
        setTransferActionEnabled(false)
    }

    override func sipCallEarly(_ notification: Notification) {
        super.sipCallEarly(notification)
        setTransferActionEnabled(false)
    }

    override func sipCallDidLocalHold(_ notification: Notification) {
        super.sipCallDidLocalHold(notification)
        callDidHoldForTransfer()
    }
}
